bash_completion

```diff
@@ -2112,32 +2112,34 @@ _included_ssh_config_files()
     configfile=$1
 
     local -a included
-    _comp_split included "$(command sed -ne 's/^[[:blank:]]*[Ii][Nn][Cc][Ll][Uu][Dd][Ee][[:blank:]]\(.*\)$/\1/p' "${configfile}")" || return
-
-    for i in "${included[@]}"; do
-        # Check the origin of $configfile to complete relative included paths on included
-        # files according to ssh_config(5):
-        #  "[...] Files without absolute paths are assumed to be in ~/.ssh if included in a user
-        #   configuration file or /etc/ssh if included from the system configuration file.[...]"
-        if ! [[ $i =~ ^\~.*|^\/.* ]]; then
-            if [[ $configfile =~ ^\/etc\/ssh.* ]]; then
-                i="/etc/ssh/$i"
-            else
-                i="$HOME/.ssh/$i"
-            fi
-        fi
-        __expand_tilde_by_ref i
-        # In case the expanded variable contains multiple paths
-        _comp_expand_glob files '$i'
-        if ((${#files[@]})); then
-            for f in "${files[@]}"; do
-                if [[ -r $f && ! -d $f ]]; then
-                    config+=("$f")
-                    # The Included file is processed to look for Included files in itself
-                    _included_ssh_config_files "$f"
+    local -a included_raw
+    local -a included_deeper
+    included=("$configfile")
+
+    while (("${#included[@]}" > 0)); do
+        included_deeper=()
+        _comp_split included_raw "$(command sed -ne 's/^[[:blank:]]*[Ii][Nn][Cc][Ll][Uu][Dd][Ee][[:blank:]]\(.*\)$/\1/p' "${included[@]}")" || return
+        for i in "${included_raw[@]}"; do
+            if ! [[ $i =~ ^\~.*|^\/.* ]]; then
+                if [[ $configfile =~ ^\/etc\/ssh.* ]]; then
```

Comment:
Shouldn't this `configfile` be replaced by the filename where the line `$i` originates from?

Reply:
Counterintuitively, no, it's shouldn't. Relative `Include` is not resolved relatively to the including file. From the manual:

```
Files without absolute paths are assumed to be in ~/.ssh if included in a user configuration file
or /etc/ssh if included from the system configuration file.
```

And here's the supporting source code: https://github.com/openssh/openssh-portable/blob/master/readconf.c#L1878

So in fact the original implementation of `_included_ssh_config_files` was buggy since it would have attempted to resolve relative Include's in the wrong place.

Reply:
> Relative `Include` is not resolved relatively to the including file.

`configfile` here is not used to resolve the relative path.

> ```
> Files without absolute paths are assumed to be in ~/.ssh if included in a user configuration file
> or /etc/ssh if included from the system configuration file.
> ```

The original code seems to try to implement this by assuming that the files starting with /etc/ssh is the system configuration file and the others are the user configuration files, though I'm not sure if that assumption is correct.

But the above description in the manual is somewhat ambiguous. Do you mean that e.g. a relative `Include` directive in a system configuration file finds a file in `~/.ssh` if the system configuration file is originally included by another `Include` in a user configuration file? That's what this PR implements. Hmm, at a quick glance in `readconf.c`, it seems to be the case, though I haven't carefully read it.

Reply:
>> Relative Include is not resolved relatively to the including file.
> configfile here is not used to resolve the relative path to the including file.
Yes, but it is used to determine if the file is resolved relatively to /etc/ssh or ~/.ssh.

> Do you mean that e.g. a relative Include directive in a system configuration file finds a file in ~/.ssh if the system configuration file is originally included by another Include in a user configuration file?
Yes. The `flags` variable is set only once, at the top level, in ssh.c:567, and SSHCONF_USERCONF bit is retained when recursing.

Reply:
Thank you for the confirmation!

This `ssh` behavior is tricky, so I think we need to leave comments in a code comment and in the commit message. Then, the main subject of the commit becomes like `fix(ssh): resolve relative ssh_config Include correctly` instead of `perf(ssh): optimize traversing ssh config includes` (Or you may separate the commit into two).